### PR TITLE
feat: Add support to convert a Query to a BundledQuery.

### PR DESCRIPTION
**Pull request by wu-hui**
Note the tricky thing here is: we introduced limitToLast, which means we need to decide if the `structuredQuery` saved in the bundle should be translated or not.

This implementation saves it without translation, which means the client SDK will use the `structureQuery` and `limitType` saved here to construct a new client `Query`, and leaves the translation to the existing logic on the client side.

### CODE REVIEW

File: dev/src/reference.ts [hunk 1]
```diff
@@ -1977,26 +1950,89 @@ export class Query<T = DocumentData> {
...
-          o.toProto()
-        );
+    }
+
+    const reqOpts: api.IRunQueryRequest = {
+      parent: parentPath.formattedName,
+      structuredQuery,
+    };
+
+    reqOpts.transaction = transactionId;
```
Comment (schmidt-sebastian) [inline dev/src/reference.ts:1960]: Nit: Merge this into the assignment above.

File: dev/src/reference.ts [hunk 2]
```diff
@@ -1977,26 +1950,89 @@ export class Query<T = DocumentData> {
...
+
+    return reqOpts;
```
Comment (schmidt-sebastian) [inline dev/src/reference.ts:1962]: Possible inline?

File: dev/src/reference.ts [hunk 3]
```diff
@@ -1977,26 +1950,89 @@ export class Query<T = DocumentData> {
...
+  }
+
+  /**
+   * Internal method for serializing a query to its RunQuery proto
+   * representation with an optional transaction id.
```
Comment (schmidt-sebastian) [inline dev/src/reference.ts:1967]: This documentation seems not applicable.

File: dev/src/reference.ts [hunk 4]
```diff
@@ -1977,26 +1950,89 @@ export class Query<T = DocumentData> {
...
+
+  private toStructuredQuery(): api.IStructuredQuery {
+    const structuredQuery: api.IStructuredQuery = {
+      from: [
+        {
+          collectionId: this._queryOptions.collectionId,
+        },
+      ],
+    };
+    if (this._queryOptions.allDescendants) {
```
Comment (schmidt-sebastian) [inline dev/src/reference.ts:2001]: Nit: Empty line above

File: dev/src/reference.ts [hunk 5]
```diff
@@ -1977,26 +1950,89 @@ export class Query<T = DocumentData> {
...
+        },
+      };
     }
 
+    if (this._queryOptions.hasFieldOrders()) {
+      structuredQuery.orderBy = this._queryOptions.fieldOrders.map(o =>
+        o.toProto()
+      );
+    }
+    structuredQuery.startAt = this.toCursor(this._queryOptions.startAt);
```
Comment (schmidt-sebastian) [inline dev/src/reference.ts:2025]: Nit: Empty line above

File: dev/src/reference.ts (hunk 1 above)
Comment (wu-hui) [reply dev/src/reference.ts:1960]: Done.

File: dev/src/reference.ts (hunk 2 above)
Comment (wu-hui) [reply dev/src/reference.ts:1962]: Done.

File: dev/src/reference.ts (hunk 3 above)
Comment (wu-hui) [reply dev/src/reference.ts:1967]: Fixed.

File: dev/src/reference.ts (hunk 4 above)
Comment (wu-hui) [reply dev/src/reference.ts:2001]: Done.

File: dev/src/reference.ts (hunk 5 above)
Comment (wu-hui) [reply dev/src/reference.ts:2025]: Done.

**Pull request merged**